hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/opt/physical/index/IndexResolver.java

```diff
@@ -135,21 +159,122 @@ private IndexResolver() {
             return Collections.emptyList();
         }
 
-        List<RelNode> rels = new ArrayList<>(supportedIndexes.size());
 
         for (MapTableIndex index : supportedIndexes) {
             // Create index scan based on candidates, if possible. Candidates could be merged into more complex
             // filters whenever possible.
-            RelNode rel = createIndexScan(scan, distribution, index, conjunctions, candidates);
 
-            if (rel != null) {
-                rels.add(rel);
+            RelNode relAscending = createIndexScan(scan, distribution, index, conjunctions, candidates, false);
+            RelNode relDescending = createIndexScan(scan, distribution, index, conjunctions, candidates, true);
+
+            if (relAscending != null) {
+                rels.add(relAscending);
+            }
+
+            if (relDescending != null) {
+                rels.add(relDescending);
             }
         }
 
         return rels;
     }
 
+
+    /**
+     * Filters out index scans which collation is covered (prefix based) by another index scan in the rels.
+     *
+     * @param rels the list of index scans
+     * @return the list of filtered out index scans
+     */
+    @SuppressWarnings("checkstyle:NPathComplexity")
+    private static List<RelNode> excludeCoveredCollations(List<RelNode> rels) {
+        // Order the index scans based on their collation
+        TreeMap<RelCollation, RelNode> relsTreeMap = new TreeMap<>((coll1, coll2) -> {
```

Comment:
I would suggest moving the comparison logic into a separate class and tests it extensively. Otherwise, we cannot test it to a sufficient extent.

Reply:
Added in https://github.com/hazelcast/hazelcast/pull/18013/commits/c4b7e90df6cdb2bca15b0c6c0976c4e9560a6411